File: Stage24/Java/Problem6549.java
package Stage24.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.StringTokenizer;

public class Problem6549 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        while (true) {
            String line = br.readLine();
            if (line.equals("0")) break;

            StringTokenizer st = new StringTokenizer(line);
            int n = Integer.parseInt(st.nextToken());
    
            long[] h = new long[n];
            for (int i = 0; i < n; i++) {
                h[i] = Long.parseLong(st.nextToken());
            }

            Stack<Integer> stack = new Stack<>();
            long max = 0;

            for (int i = 0; i < n; i++) {
                while(!stack.isEmpty() && h[stack.peek()] > h[i]) {
                    int heightIdx = stack.pop();
                    long height = h[heightIdx];

                    int width;
                    if (stack.isEmpty()) {
                        width = i;
                    } else {
                        width = i - stack.peek() - 1;
                    }

                    max = Math.max(max, height * width);
                }
                stack.push(i);
            }

            while (!stack.isEmpty()) {
                int heightIdx = stack.pop();
                long height = h[heightIdx];

                int width;
                if (stack.isEmpty()) {
                    width = n;
                } else {
                    width = n - stack.peek() - 1;
                }

                max = Math.max(max, height * width);
            }

            sb.append(max).append('\n');
        }

        System.out.println(sb);
        br.close();
    }
}

File: Stage24/Python/Problem6549.py
import sys
input = sys.stdin.readline

def main():
    while True:
        line = input().split()
        if not line:
            break

        n = int(line[0])
        if n == 0:
            break

        h = list(map(int, line[1:]))

        stack = []
        max_area = 0

        for i in range(n):
            while stack and h[stack[-1]] > h[i]:
                height_idx = stack.pop()
                height = h[height_idx]

                if not stack:
                    width = i
                else:
                    width = i - stack[-1] - 1

                max_area = max(max_area, height * width)

            stack.append(i)

        while stack:
            height_idx = stack.pop()
            height = h[height_idx]

            if not stack:
                width = n
            else:
                width = n - stack[-1] - 1

            max_area = max(max_area, height * width)

        print(max_area)
        
if __name__ == "__main__":
    main()